src/editors/matl.rs

```diff
@@ -592,6 +610,37 @@ fn menu_bar(
             }
         });
 
+        ui.menu_button("Reorder", |ui| {
+            if ui.button("Move Material to Top").clicked() {
+                ui.close_menu();
+                matl.entries.swap(0, state.selected_material_index);
+                state.selected_material_index = 0;
+                changed = true;
+            }
+
+            if ui.button("Move Material to Bottom").clicked() {
+                ui.close_menu();
+                let last_index = matl.entries.len() - 1;
+                matl.entries.swap(state.selected_material_index, last_index);
+                state.selected_material_index = last_index;
+                changed = true;
+            }
+            
+            ui.separator();
+
+            if ui.button("Sort Materials").clicked() {
+                ui.close_menu();
+                matl.entries.sort_by_key(|k| k.material_label.to_lowercase().clone());
+                for (material_index, material) in matl.entries.iter().enumerate() {
+                    if matl.entries[state.selected_material_index].material_label == material.material_label {
```

Comment:
`state.selected_index` is no long referencing the same material after sorting. You'll need to store information about the selected material before sorting. I think it would be a little more idiomatic to use iterators instead of a for loop to set the index. https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.position

Reply:
I believe the clone after to_lowercase is redundant. `cargo clippy` should tell you.

Reply:
```error[E0507]: cannot move out of a shared reference
   --> src\editors\matl.rs:634:81
    |
634 |                 let new_selected_material_index = matl.entries.iter().position(|&x| x == matl.entries[state.selected_material_index]);
    |                                                                                 ^-
    |                                                                                  |
    |                                                                                  data moved here
    |                                                                                  move occurs because `x` has type `ssbh_data::matl_data::MatlEntryData`, which does not implement the `Copy` trait
    |
help: consider removing the borrow```

Reply:
`cargo clippy` gave no error regarding the clone.

Reply:
It should still work without the clone call since to_lowercase returns String instead of &str. Comparing with the entrie selected material entry will work, but you'll need to clone it before sorting to avoid borrowing issues. A reference wouldn't point to the same entry after sorting. It's similar to why the selected index needs to be changed after sorting.

Reply:
I don't think the position function works without using a reference, which is making it not work.

Reply:
You shouldn't need a reference on the variable x in the closure. The call to iter already gives you a reference if you inspect the type of x.

Reply:
```error[E0277]: can't compare `&ssbh_data::matl_data::MatlEntryData` with `ssbh_data::matl_data::MatlEntryData`
   --> src\editors\matl.rs:634:86
    |
634 |                 let new_selected_material_index = matl.entries.iter().position(|x| x == matl.entries[state.selected_material_index]);
    |                                                                                      ^^ no implementation for `&ssbh_data::matl_data::MatlEntryData == ssbh_data::matl_data::MatlEntryData`
    |```

Reply:
It's complaining that the two sides are different types. You can just add a reference to the owned one to make the compiler happy. You should be able to do something like this.
```rust
let selected_material = matl.entries[state.selected_material_index].clone();
matl.entries.sort_by(...);
state.selected_material_index = matl.entries.iter.position(|e| e == &selected_material).unwrap_or_default();
```